csharp/src/Ice/ConnectionI.cs

```diff
@@ -607,15 +607,23 @@ public override bool startAsync(int operation, Ice.Internal.AsyncCallback comple
                             observerStartWrite(_writeStream.getBuffer());
                         }
 
-                        bool completed;
-                        if (_transceiver.startWrite(_writeStream.getBuffer(), completedCallback, this, out completed))
+                        bool messageWritten = false;
+                        bool completedSynchronously =
+                            _transceiver.startWrite(
+                                _writeStream.getBuffer(),
+                                completedCallback,
+                                this,
+                                out messageWritten);
+                        // If the startWrite call consumed the complete buffer, we assume the message is sent now for
+                        // at-most-once semantics.
+                        if (messageWritten && _sendStreams.Count > 0)
```

Comment:
(Not from this PR)
Do you know when we can reach this point with _sendStreams.Count == 0?

Reply:
Seems `StartWrite` is some times called with a 0 buffer size just to invoke the completion callback. It is not clear why we do it that way, and cannot call it from the finishAsync.